File: cxx/traaxx/sources/propagate_down_pointer_doubling.cpp
#include <traaxx/propagate_down_pointer_doubling.hpp>
#include <traaxx/level_count_for.hpp>
#include <algorithm>
#include <execution>
#include <numeric>

namespace traaxx
{
    template<typename IndexT>
    BitMask propagate_down_pointer_doubling(const std::vector<IndexT> &parent, BitMask seed, IndexT d_max)
    {
        auto const n = parent.size();
        if (n == 0)
        {
            return seed;
        }
        auto indices = std::vector<IndexT>(n);
        std::iota(indices.begin(), indices.end(), IndexT{ 0 });
        auto const level_count = level_count_for(d_max);
        auto current = std::move(seed);
        auto anc = parent;
        for (IndexT k = 0; k < level_count; ++k)
        {
            auto next = BitMask(n);
            std::for_each(
                std::execution::par,   //
                indices.cbegin(),      //
                indices.cend(),        //
                [&](IndexT i)
                {
                    if (current.get(i) || current.get(anc[i]))
                    {
                        next.atomic_or(i);
                    }
                });
            current = std::move(next);
            if (k + IndexT{ 1 } < level_count)
            {
                auto next_anc = std::vector<IndexT>(n);
                std::transform(
                    std::execution::par,   //
                    indices.cbegin(),      //
                    indices.cend(),        //
                    next_anc.begin(),      //
                    [&](IndexT i) { return anc[anc[i]]; });
                anc = std::move(next_anc);
            }
        }
        return current;
    }

    template BitMask propagate_down_pointer_doubling<std::uint32_t>(
        const std::vector<std::uint32_t> &, BitMask, std::uint32_t);
}

File: go/std/propagate_down_pointer_doubling.go
package std

// PropagateDownPointerDoubling propagates a mask downward in O(log d) steps
// via successively doubled ancestor pointers (A1, A2, A4, ...), rather than
// iterating to a detected fixed point: the number of steps (level_count) is
// fixed up front from dMax. Mirrors traaxx::propagate_down_pointer_doubling
// (cxx/traaxx/sources/propagate_down_pointer_doubling.cpp).
func PropagateDownPointerDoubling(parent []Index, seed BitMask, dMax Index) BitMask {
	n := len(parent)
	if n == 0 {
		return seed
	}
	levelCount := int(LevelCountFor(dMax))
	current := seed
	anc := make([]Index, n)
	copy(anc, parent)
	for k := 0; k < levelCount; k++ {
		next := NewBitMask(n)
		parallelFor(n, func(lo, hi int) {
			for i := lo; i < hi; i++ {
				if current.Get(Index(i)) || current.Get(anc[i]) {
					next.AtomicOr(Index(i))
				}
			}
		})
		current = next
		if k+1 < levelCount {
			nextAnc := make([]Index, n)
			parallelFor(n, func(lo, hi int) {
				for i := lo; i < hi; i++ {
					nextAnc[i] = anc[anc[i]]
				}
			})
			anc = nextAnc
		}
	}
	return current
}
